File: leetcode_0051/cpp/leetcode_0051.cpp
//
// Created by wzy on 2022/4/19.
//

#include<iostream>
#include<queue>
#include<vector>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <list>
#include <deque>
#include <stack>
#include <sstream>
#include <typeinfo>
#include <climits>
#include <random>
#include <ctime>

using namespace std;


class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
        //思路:使用字符串数组来表示每一行 或者 int数组来表示每一行皇后的位置
        //回溯每一行，对每一行的每一列摆放进行验证
        vector<string> vec = vector<string>(n,string(n,'.'));
        vector<vector<string>> res;
        dfs(res,vec,n,0);
        return res;
    }
private:
    void dfs(vector<vector<string>>& res,vector<string>& vec,int n,int row){
        if(row == n){
            //说明前row行进行了皇后摆放的插入
            res.push_back(vec);
            return ;
        }
        for(int i = 0;i < n;i++){
            if(valid(vec,n,row,i)){
                vec[row][i] = 'Q';
                dfs(res,vec,n,row+1);
                vec[row][i] = '.';
            }
        }
    }
    bool valid(vector<string>& vec,int n,int row,int idx){
        //判断是否在同一列
        for(int i = 0;i < row;i++)
            //说明有在同一列
            if(vec[i][idx] == 'Q')
                return false;
        //判断是否在左上方
        // row - idx = i - x
        for(int i = 0;i < row;i++){
            int j = i - row + idx;
            if( j >= 0 && j < n && vec[i][j] == 'Q')
                return false;
        }
        //判断是否在右上方
        // i - (n-x) = row - (n-idx)
        for(int i = 0;i < row;i++){
            int j = row+idx-i;
            if( j >= 0 && j < n && vec[i][j] == 'Q')
                return false;
        }
        return true;
    }
};

template<typename T>
string convertVecToStr(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    return res;
}

template<typename T>
void printVecArr(std::vector<std::vector<T>> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << convertVecToStr(ele);
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

template<typename T>
void printVec(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

int main(){
    int n = 4;
    auto res = Solution().solveNQueens(n);
    printVecArr(res);
    //result
    //[[.Q..,...Q,Q...,..Q.],[..Q.,Q...,...Q,.Q..]]
}

File: leetcode_0051/java/leetcode_0051.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class leetcode_0051 {
    class Solution {
        public List<List<String>> solveNQueens(int n) {
            //思路:使用字符串数组来表示每一行 或者 int数组来表示每一行皇后的位置
            //回溯每一行，对每一行的每一列摆放进行验证
            char[] chs = new char[n];
            Arrays.fill(chs,'.');
            List<StringBuilder> sub = new ArrayList(n);
            for(int i = 0;i < n;i++)
                sub.add(new StringBuilder(String.valueOf(chs)));
            List<List<String>> res = new LinkedList();
            dfs(res,sub,n,0);
            return res;
        }
        private void dfs(List<List<String>> res,List<StringBuilder> sub,int n,int row){
            if(row == n){
                LinkedList<String> res_sub = new LinkedList();
                for(int i = 0;i < n;i++)
                    res_sub.add(sub.get(i).toString());
                res.add(res_sub);
                return ;
            }
            for(int i = 0;i < n;i++){
                if(valid(sub,n,row,i)){
                    sub.get(row).setCharAt(i,'Q');
                    dfs(res,sub,n,row+1);
                    sub.get(row).setCharAt(i,'.');
                }
            }
        }
        private boolean valid(List<StringBuilder> sub,int n,int row,int idx){
            //判断是否在同一列
            for(int i = 0;i < row;i++){
                if(sub.get(i).charAt(idx) == 'Q')
                    return false;
            }
            //判断是否在左上方
            for(int i = 0;i < row;i++){
                //row - idx = i - x
                int j = i - row + idx;
                if(j >= 0  && j < n && sub.get(i).charAt(j) == 'Q')
                    return false;
            }
            //判断是否在右上方
            for(int i = 0;i < row;i++){
                //row - (n-idx) = i - (n-x)
                int j = row + idx - i;
                if(j >= 0 && j < n && sub.get(i).charAt(j) == 'Q')
                    return false;
            }
            return true;
        }
        public <T> void printListArr(List<List<T>> list) {
            for (int i = 0; i < list.size(); i++) {
                if (i == list.size() - 1) {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.println(Arrays.toString(list.get(i).toArray((T[]) new Object[0])) + " ] ");
                } else {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.printf(Arrays.toString(list.get(i).toArray((T[]) new Object[0])) + " , ");
                }
            }
        }
    }
    public static void main(String[] args) {
        int n = 4;
        leetcode_0051.Solution solution = new leetcode_0051().new Solution();
        List<List<String>> res = solution.solveNQueens(n);
        solution.printListArr(res);
        //result
        //[ [.Q.., ...Q, Q..., ..Q.] , [..Q., Q..., ...Q, .Q..] ]
    }
}
